candle-nn/src/encoding.rs

```
@@ -0,0 +1,193 @@
+/// By default `on_value` is `1.0` and `off_value` is `0.0`.
+///
+/// Other encoding values can be used by setting `on_value` and `off_value` to the desired values.
+///
+/// # Examples
+///
+///```rust
+/// use candle::{Shape, Tensor, Device};
+/// use candle_nn::encoding::one_hot;
+///
+/// let device = Device::Cpu;
+/// let depth = 4;
+/// let indices = Tensor::new(vec![vec![0f32, 2.], vec![1., -1.]], &device).unwrap();
+///
+/// let on_value = Some(1.0); // default
+/// let off_value = Some(0.0); // default
+///
+/// let one_hot = one_hot::<f32>(indices, depth, on_value, off_value).unwrap();
+/// let expected_matrix = [
+///     [[1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0]],
+///     [[0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0]],
+/// ];
+/// assert_eq!(one_hot.shape(), &Shape::from((2, 2, 4)));
+///
+/// let matrix = one_hot.to_vec3::<f32>().unwrap();
+///
+/// assert_eq!(matrix, expected_matrix);
+/// ```
+///
+/// # Bails
+///
+/// This method will bail on tensors with a rank greater than 3.
+///
+/// # API Design
+///
+/// The api design for this method is loosely based on the [TensorFlow One-Hot](https://www.tensorflow.org/api_docs/python/tf/one_hot) method.
+pub fn one_hot<D: WithDType>(
+    indices: Tensor,
+    depth: usize,
+    on_value: Option<D>,
+    off_value: Option<D>,
+) -> Result<Tensor> {
+    let on_value = on_value.unwrap_or(D::from_f64(1.0));
+    let off_value = off_value.unwrap_or(D::from_f64(0.0));
+
+    let rank = indices.rank();
+    let dims = indices.dims();
+    match rank {
+        0 => {
+            let mut v = vec![off_value; depth];
+            let index = indices.to_vec0::<f32>()?;
+            v[index as usize] = on_value;
+
+            Tensor::new(v, indices.device())
+        }
+        1 => {
+            let mut v = vec![off_value; depth * dims[0]];
+            let index = indices.to_vec1::<f32>()?;
+            for i in 0..dims[0] {
+                if index[i] < 0.0 {
```

These errors should be caught and sent back to the user via some `bail` rather than silently swallowed.